### DISCUSSION TITLE
[Question] Adding new inventory fields sourced from products database

### DISCUSSION THREAD

**Discussion by TheVirtualMoose** [YetiForceCompany/YetiForceCRM#14800] [How to]
I am trying to add new inventory fields that will take their values from appropriate values in the products' database. So far, I have figured out I need to:

- Create a new class in (basedir)/modules/Vtiger/inventoryfields
- Add a corresponding .tbl snippet in (basedir)/layouts/basic/modules/Vtiger/inventoryfields
- Add a setField function to (basedir)/public_html/layouts/basic/modules/Vtiger/resources/Inventory.js and call it from setRowData

This works if I merely calculate the value from values already present in the inventory row, but I cannot get the new inventory field to get its value from the products' database. I tried mimicking Unit and SubUnit fields but with little success.

Any documentation on how to do that or even pointers to files I should take a look at would be very welcome.

**Comment by TheVirtualMoose** [comment]
OK, after fiddling with that a bit more, I discovered I messed up the field in Products module I was trying to map to. Having fixed that, I found it easy to use autoFields with these steps:
- Create a new inventory field (tested with Value fields only so far)
- (Optional) rename the field in the relevant _inventory and _invfield tables (column name in _inventory, value in 'columnname' field of relevant row in _invmap).
- Add mapping in the relevant _invmap table. NOTE: "field" denotes field name in the source module (e.g. Products), "tofield" denotes field name in the inventory.

If a new field type is needed, the steps above are good but need to be expanded:
- Create a new class in (basedir)/modules/Vtiger/inventoryfields
- Add a corresponding .tbl snippet in (basedir)/layouts/basic/modules/Vtiger/inventoryfields
- Add a setField function to (basedir)/public_html/layouts/basic/modules/Vtiger/resources/Inventory.js and call it from setRowData **and from mapResultsToFields**

I hope this is useful to others. It would be great to have these features better documented, because right now the only way to figure them out is by trial-and-error and by reading code.

**Reply by vovpff** [reply]
> * Add a setField function to (basedir)/public_html/layouts/basic/modules/Vtiger/resources/Inventory.js and call it from setRowData **and from mapResultsToFields**

Could you share your code for setField function?

**Reply by apcloic** [reply]
> > * Add a setField function to (basedir)/public_html/layouts/basic/modules/Vtiger/resources/Inventory.js and call it from setRowData **and from mapResultsToFields**
> 
> Could you share your code for setField function?

+1 I'm interested too.
Ty

**Reply by TheVirtualMoose** [reply]
> > * Add a setField function to (basedir)/public_html/layouts/basic/modules/Vtiger/resources/Inventory.js and call it from setRowData **and from mapResultsToFields**
> 
> Could you share your code for setField function?

It's rather simple, a basic function would look like that:
setField(row, val) {
	val = App.Fields.Double.formatToDisplay(val); //Adjust to the field's type. AFAIK this line is not needed for strings
	row.find('.field').val(val);
	row.find('.fieldText').text(val);
},

You can take a look at my modifications to Inventory.js to figure out how this goes with getField() and calculateField() function - look at fragments enclose in //Custom and //End custom coments. Caveat, I am not a coder and definitely not a JS coder, and what I did is probably a coding atrocity, but it does what I need.
[Inventory.zip](https://github.com/YetiForceCompany/YetiForceCRM/files/6647152/Inventory.zip)

**Comment by newminder** [comment]
did you solve the problem?  I happened the same problem, and still stuck there.
